### ISSUE TITLE
Build general purpose two-tiered component shutdown mechanism

### ISSUE THREAD

**Issue by mihaitodor** [Jeffail/benthos#688]
While enabling the staticcheck linter in #687 I noticed that there are some potential concurrency issues in [lib/input/socket_server.go](https://github.com/Jeffail/benthos/blob/44ccb43320990c9e0e61e75c63dfd820f1fb0780/lib/input/socket_server.go) namely:

- ~[SocketServer.retriesMut](https://github.com/Jeffail/benthos/blob/44ccb43320990c9e0e61e75c63dfd820f1fb0780/lib/input/socket_server.go#L92) gets locked and unlocked a bunch of times, but without effect. Even that `hasLocked` bool in `sendMsg` is never set to true.~ - Turns out this is a back pressure mechanism (thanks Ash!). The `hasLocked` flag issue has been addressed in #687.
- In the [current code](https://github.com/Jeffail/benthos/blob/44ccb43320990c9e0e61e75c63dfd820f1fb0780/lib/input/socket_server.go#L243-L254), there seems to be a race on calling `t.listener.Close()` from both the goroutine and the defer statement.
- [Cache-ing a context](https://github.com/Jeffail/benthos/blob/44ccb43320990c9e0e61e75c63dfd820f1fb0780/lib/input/socket_server.go#L95) in a struct is considered an antipattern: https://blog.golang.org/context-and-structs I guess in this case it kinda' makes sense to do so, since it's internal to the SocketServer, but perhaps a quit chan would be more idiomatic?

**Comment by Jeffail** [comment]
Nice, added two comments, I think I need to dig into the lock stuff as we definitely need it, not sure why it's being flagged but it's probably something I've missed out in my implementation somewhere.

I saw the context post, I'm not opposed to replacing it with something else. In fact there are some extra features I'd like to have for a component shutdown mechanism that context doesn't have. However, I got to this point because I was originally using a combination of atomics, `sync.Once` and channels and it was a total mess.

The main two _contexts_ :smirk: where `context` helps out quite a bit:

## Close a channel definitely only once

Before:

```go
w.closeOnce.Once(func() {
  close(w.closeChan)
})
```

With `context`:

```go
w.close()
```

With a hypothetical new component:

```go
w.shutdown.AtLeisure()
```

## Check if channel is closed

Before:

```go
closed := false
select {
  case <-w.closeChan:
    closed = true
  default:
}
```

With `context`:

```go
closed := w.ctx.Err()
```

With a hypothetical new component:

```go
closed := w.shutdown.Triggered()
```

Most components within Benthos have two tiers of shutdown request. The first is to close "at your leisure", which means if you're in the middle of something then go ahead and finish up, this would include flushing any pending messages, waiting for acknowledgements, etc. The second is to shut down ASAP because you're holding up something important (we've waited for too long to exit the process, I want to replace you with something else, etc).

A lot of components therefore need two shutdown signals which means a complex cocktail of sync types becomes double trouble to deal with. Since all of this stuff is nicely baked into `context` it's an obvious thing to reach for, at the cost of being a little confusing since it's being adopted for a purpose it's not traditionally meant for, which I'll admit is especially hazardous when we have methods that actually use context parameters in the proper way.

I'll open a new issue for the socket lock stuff if there's more work to do and repurpose this one for building a new shutdown mechanism.

**Comment by mihaitodor** [comment]
Thank you so much for this detailed explanation! I learned some new things and I'm curious how this will look like once implemented. It sounds tricky to get right...

**Comment by Jeffail** [comment]
Here's a quick attempt at a shutdown signalling mechanism: https://github.com/Jeffail/benthos/commit/3e427023e014dd974d5f5ef745e80e3209e7e324, going to test out the API with the `aws_sqs` refactor.

**Comment by mihaitodor** [comment]
@Jeffail Neat! I was thinking this will require some sort of state machine model, which is what I think you aimed for too.

I'm guessing the `Close*Chan()` methods are available to avoid creating extra context objects by calling the `Close*Ctx()` methods when that's not necessary.

**Comment by Jeffail** [comment]
@mihaitodor whoops forgot to revisit this one. Yeah exactly, we basically have a super simple state machine:

```
open ---> closing at leisure ---> closing now ---> closed
```

And we can skip straight to any step, but can't go backwards. The inside of a goroutine following this shutdown mechanism is a read-only consumer of the states `closing at leisure` and `closing now`, which is triggered from the outside, and the outside is a read-only consumer of the state `closed`, which is triggered from the inside.

The API itself doesn't enforce the inside/outside restrictions so it's up to the component implementation to respect that.

For the most part the signals are being consumed as a channel read and therefore we don't need to create a cancellable context for each API call. When contexts are required the API allows you to wrap an existing context with signal cancellation rather than creating a new one.

I've tried it out in a few different components already and it seems to fit the bill. I haven't eliminated owned contexts everywhere yet but the general rule is that when a component is updated and isn't using this new shutdown signaller then it should be migrated at the same time.

**Comment by mihaitodor** [comment]
Awesome, that makes a lot of sense @Jeffail. Thank you for the detailed explanation! :)
